exp/views/responses.py

```diff
@@ -1037,114 +901,498 @@ def build_framedata_dict_csv(writer, responses):
+            else "",
+            extension,
+        )
+
+        if data_type == "json":
+            cleaned_data = json.dumps(
+                construct_response_dictionary(resp, RESPONSE_COLUMNS, header_options),
+                indent="\t",
+                default=str,
+            )
+        elif data_type == "csv":
+            row_data = flatten_dict(
+                {col.id: col.extractor(resp) for col in RESPONSE_COLUMNS}
+            )
+            header_list = get_response_headers(header_options, row_data.keys())
+            output, writer = csv_dict_output_and_writer(header_list)
+            writer.writerow(row_data)
+            cleaned_data = output.getvalue()
+        elif data_type == "framedata":
+            cleaned_data = build_single_response_framedata_csv(resp)
+        response = HttpResponse(cleaned_data, content_type="text/{}".format(extension))
+        response["Content-Disposition"] = 'attachment; filename="{}"'.format(filename)
+        return response
+
+
+class StudyResponseVideoAttachment(
+    ExperimenterLoginRequiredMixin, UserPassesTestMixin, StudyLookupMixin, View
+):
     """
-    Builds CSV file contents for overview of all responses
+    View that redirects to a requested video for a study response.
     """
 
-    headers = set()
-    session_list = []
+    raise_exception = True
+    video = None
+
+    def get_video(self):
+        # Only select the video from consented videos for this study
+        if self.video is None:
+            self.video = self.get_study().videos_for_consented_responses.get(
+                pk=self.kwargs.get("video")
+            )
+        return self.video
 
-    paginator = Paginator(responses, RESPONSE_PAGE_SIZE)
-    for page_num in paginator.page_range:
-        page_of_responses = paginator.page(page_num)
-        for resp in page_of_responses:
-            row_data = get_response_headers_and_row_data(resp)["dict"]
-            # Add any new headers from this session
-            headers = headers | set(row_data.keys())
-            session_list.append(row_data)
-
-    header_list = get_response_headers(optional_headers_selected_ids, headers)
-    output, writer = csv_dict_output_and_writer(header_list)
-    writer.writerows(session_list)
-    return output.getvalue()
+    def can_view_this_video(self):
+        user = self.request.user
+        study = self.get_study()
+        video = self.get_video()
+
+        return user.is_researcher and (
+            (
+                user.has_study_perms(StudyPermission.READ_STUDY_RESPONSE_DATA, study)
+                and not video.response.is_preview
+            )
+            or (
+                user.has_study_perms(StudyPermission.READ_STUDY_PREVIEW_DATA, study)
+                and video.response.is_preview
+            )
+        )
 
+    test_func = can_view_this_video
 
-def build_single_response_framedata_csv(response):
+    def post(self, request, *args, **kwargs):
```

Comment:
This isn't really RESTful - POST is for [creating resources on the server](https://docs.python.org/3/tutorial/datastructures.html#dictionaries); you should change the template to use an anchor tag and just use regular GET and refactor accordingly

Reply:
darnit, thanks, I have an incorrect association about POST and "submitting form data" even though this is clearly just... what is being requested. Fixed to just use GET method, but StudyResponsesList also uses post to handle requests for individual-response json/csv files with specified headers. Should that be a separate individual-download view using GET?

Reply:
Yeah, I would go ahead and break out a separate view for that. There's kind of another option with Content Negotiation but this is not really what that's for.

Reply:
ok, this is its own view now.